### ISSUE TITLE
Support Outlook.com Client Credentials Flow

### ISSUE THREAD

**Issue by hannibal218bc** [tarickb/sasl-xoauth2#76]
Hi,

I'd like to use this project to authenticate as an application to a Microsoft365 tenant, but it seems that's not supported?

There is `client_id`, `client_secret` and `token_endpoint` (`https://login.microsoftonline.com/<TENANT>/oauth2/v2.0/token`) provided in `/etc/sasl-xoauth2.conf`. When attempting to send mail, postfix logs:

> sasl-xoauth2: auth failed: 2023-10-11 00:18:22: TokenStore::Read: missing refresh_token

However, I can successfully use Postman on these exact values to retrieve an `access_token`. The client credentials flow should directly use this token as long as it is valid...

The credentials belong to an "app registration" and already has all necessary rights and consents, hence I don't want to go through any additional consent dialog (as the `--use-device-flow` flow would suggest).

Did I miss something? Thank you very much!

**Comment by tarickb** [comment]
Yes, authenticating with Microsoft 365 is supported. Please follow the instructions in the [README](https://github.com/tarickb/sasl-xoauth2#outlookoffice-365-configuration-legacy-client-deprecated), including the use of `sasl-xoauth2-tool`. I'm not familiar with Postman or with what specific use case you have but sasl-xoauth2 will not work without a refresh token (how else would it obtain an access token when the access token expires?).

**Comment by hannibal218bc** [comment]
Yes, that's the point: with the "client credentials" flow, there really is no refresh token. You use the credentials to directly obtain an access token, which is then valid for a certain time. After that, you fetch a new access token (again providing client_id and client_secret).

The README.md describes only setups where you must, at some point during setup, log in to Microsoft as a user and grant the application access on your behalf. The resulting tokens allow the application to act *as this user* .

In the flow I need to use, the application acts *on its own*. All necessary rights have been granted directly by the tenant administrator, there is no further user interaction or consent necessary.

We are successfully using these credentials with MSAL, so I'm quite sure that the setup process documented in README.md do not apply to this kind of token/flow... ? Thanks again...

**Comment by tarickb** [comment]
Ah I see -- I wasn't aware Microsoft 365 had that as an option. sasl-xoauth2 doesn't support that sort of authentication at the moment but I'll leave this open as a feature request.

**Comment by minusdavid** [comment]
I've been reading more about Microsoft's support for the client credentials flow, and it certainly seems like the best way to go when using XOAUTH2.

Google doesn't support that flow/grant type which is unfortunate. 

I'm helping some folk with XOAUTH2 support in Perl and I think it would be necessary to have configuration specifying which grant type is used so that the code knows whether to just use client credentials or whether it needs to use refresh tokens.

**Comment by minusdavid** [comment]
If I had more time and money, it would be tempting to have a crack at a PR for the client credentials flow...

**Comment by minusdavid** [comment]
It's also been a while since I've written C++ but it would be a fun passion project...

**Comment by minusdavid** [comment]
I've got the Client Credentials flow working with Microsoft using application Perl code (someone else's code but it took some work to get the Microsoft configuration right) at https://bugs.koha-community.org/bugzilla3/show_bug.cgi?id=38338

--

Technically, I suppose that you could use the existing code for sasl-xoauth2 and use client credentials to get the initial token and then use sasl-xoauth2-tool in a cronjob to fetch new access tokens every 30 minutes (prior to the 60 minute expiry I think they have with this flow).

--

In terms of a PR... I'm trying to figure out the best place in the code to change. I think it would probably be enough to update client.cc to use a different token fetcher than token_store.cc in the InitialStep, and then in the TokenSentStep to avoid trying to do a Refresh. Not sure which is the optimal place to check for the grant type, but that's all right. Easy enough to optimise that later.

**Comment by minusdavid** [comment]
I don't think I've written C++ in 10 years, so that took a bit longer than one would hope, but got there in the end. 

I've posted a PR at https://github.com/tarickb/sasl-xoauth2/pull/123

Having that working Perl code made it a lot easier to test this since I knew I already had the Microsoft side of things configured correctly. 

Please let me know if there's anything I need to change. Feel free to change whatever you like with it as well. I tried to make a very light touch.

**Comment by Steltek** [comment]
> I've been reading more about Microsoft's support for the client credentials flow, and it certainly seems like the best way to go when using XOAUTH2.

That is not true in all situations, especially with security in mind: Unlike the delegated credential flows, the client credentials flow uses application level permissions and thus bypasses any user-authentication and permissions. While this makes it simpler/easier to implement, it is granting the underlying application unfettered access to your entire environment (unless separately scoped down), thus making it extremely dangerous.

An app that has application-level permissions can apply these permissions against ANY mailbox (again, unless manually scoped down with configuration on the backend). An app with Delegated permissions will inherently only have access to the mailbox/items that the user account who delegated it access has permissions for, and that delegation has to be explicitly done through signing into that account at least once.

-> From a security perspective, delegated permissions/token flows are the best way to go, not the client credentials grant.

**Comment by minusdavid** [comment]
I don’t think that I’ve seen that in practice. 

It’s been a little while now since I configured the Microsoft account but after reviewing https://learn.microsoft.com/en-us/exchange/client-developer/legacy-protocols/how-to-authenticate-an-imap-pop-smtp-application-by-using-oauth#use-client-credentials-grant-flow-to-authenticate-smtp-imap-and-pop-connections I recall that I did need to add the mailbox to the application’s service principal. It didn’t have unfettered access. 

Could you point me at some documentation or process that creates the overpermissioned account?

**Comment by minusdavid** [comment]
Recently I’ve been using Dovecot’s OAuth protected submission service (using Keycloak for the IdP) and it works similarly. I create a service account scoped to that Oauth2 client application and use a client credentials grant which gets me access just for that user… I’d have to review the authorizations in the Dovecot config…

**Comment by minusdavid** [comment]
@Steltek Thanks for bringing this up. Ideally, I want to provide support for all flows.

**Comment by Steltek** [comment]
> It’s been a little while now since I configured the Microsoft account but after reviewing https://learn.microsoft.com/en-us/exchange/client-developer/legacy-protocols/how-to-authenticate-an-imap-pop-smtp-application-by-using-oauth#use-client-credentials-grant-flow-to-authenticate-smtp-imap-and-pop-connections I recall that I did need to add the mailbox to the application’s service principal. It didn’t have unfettered access.

This seems to be new. Apparently MS has been changing apps recently, and transitioning them to [requiring a Service Principal](https://learn.microsoft.com/en-us/entra/identity-platform/retire-service-principal-less-authentication) as of March 2026 (which would inherently bring with it the scoping to being only able to access what the SP can access).

I know for a fact that I've had to create [Application Access Policies](https://learn.microsoft.com/en-us/exchange/permissions-exo/application-access-policies) (no longer supported) in Exchange Online to reign in app access in the past. Quote from that page: `For example, the Mail.Read application permission allows apps to read mail in all mailboxes without a signed-in user.`

**Comment by minusdavid** [comment]
Ahhh thanks for sharing that. I don't use Microsoft myself, but support 1 organisation that does. Looking back it looks like I was doing my work in May, so that's why I must've fallen into the post-March 2026 world. Good to know it's changed so recently!

**Comment by minusdavid** [comment]
I was so excited to see your comments, as I thought it was a sign of life on the project, but it looks like Tarick is still MIA. 

This project inspired me to write my own non-interactive SASL plugin https://github.com/minusdavid/simple-sasl-xoauth2 so I don't think I'll do any more work on this one. With mine the SASL plugin just handles the XOAUTH2 format and relies on the application callbacks to provide the token. So with Postfix I use a socketmap lookup to use a locked down Unix socket to go to a separate process. And that separate process could be written in just about anything. I wrote a simple Perl daemon that uses a client credentials grant, but could also use a refresh token grant. I've got a little Perl bootstrapping tool kicking around as well. I'll probably do a separate repo for the token provider service.

**Comment by minusdavid** [comment]
I've also used simple-sasl-xoauth2 with Authen::SASL::XS to leverage it in Perl apps as well. Any languages with bindings for Cyrus SASL could make use of it quite readily.
